nautobot/extras/models/groups.py

```diff
@@ -0,0 +1,203 @@
+"""Dynamic Groups Models."""
+
+from django import forms
+from django.contrib.contenttypes.models import ContentType
+from django.core.serializers.json import DjangoJSONEncoder
+from django.db import models
+from django.urls import reverse
+
+from nautobot.core.models import BaseModel
+from nautobot.extras.groups import BaseDynamicGroupMap
+from nautobot.extras.models import ChangeLoggedModel
+from nautobot.utilities.utils import get_filterform_for_model, get_filterset_for_model
+from nautobot.utilities.querysets import RestrictedQuerySet
+
+
+class DynamicGroupQuerySet(RestrictedQuerySet):
+    """Queryset for `DynamicGroup` objects."""
+
+    def get_for_object(self, obj):
+        """
+        Return all `DynamicGroup` assigned to the given object.
+        """
+        if not isinstance(obj, models.Model):
+            raise TypeError(f"{obj} is not an instance of Django Model class")
+
+        # Check if dynamicgroup is supported for this model
+        model = obj._meta.model
+        dynamicgroupmap = dynamicgroup_map_factory(model)
+
+        # If there's not a map just return an unfiltered queryset?
+        # TODO(jathan): Are we sure that's what we want?
+        if not dynamicgroupmap:
+            return self
+
+        dynamicgroup_filter = dynamicgroupmap.get_queryset_filter(obj)
+        return self.filter(content_type=ContentType.objects.get_for_model(obj)).filter(dynamicgroup_filter)
+
+
+def dynamicgroup_map_factory(model):
+    """Generate a `FooDynamicGroupMap` class for a given `model`."""
+
+    filterset = get_filterset_for_model(model)
+    filterform = get_filterform_for_model(model)
+
+    group_map = type(
+        str("%sDynamicGroupMap" % model._meta.object_name),
+        (BaseDynamicGroupMap,),
+        {"model": model, "filterset": filterset, "filterform": filterform},
+    )
+
+    return group_map
+
+
+class DynamicGroup(BaseModel, ChangeLoggedModel):
+    """Dynamic Group Model."""
+
+    name = models.CharField(max_length=100, unique=True, help_text="Dynamic Group name")
+    slug = models.SlugField(max_length=100, unique=True, help_text="Unique slug")
+    description = models.CharField(max_length=200, blank=True)
+    content_type = models.ForeignKey(
+        to=ContentType,
+        on_delete=models.CASCADE,
+        verbose_name="Object Type",
+        help_text="The type of object for this group.",
+    )
+    filter = models.JSONField(
+        encoder=DjangoJSONEncoder,
+        editable=False,
```

Comment:
So that it doesn't show up in forms. It's a complex field only modified by the form internals at this time. I am not happy with this pattern right now but due to the dynamism of the form by merging in the `FooFilterForm` fields, there's not currently an easy way to move this construction logic to the model.